java-core/google-cloud-core/src/main/java/com/google/cloud/ServiceOptions.java

```diff
@@ -767,4 +791,48 @@ public String getClientLibToken() {
   public String getQuotaProjectId() {
     return quotaProjectId;
   }
+
+  /**
+   * Returns the resolved endpoint for the Service to connect to Google Cloud
+   *
+   * <p>The resolved endpoint is always in `host:port` format
+   */
+  @InternalApi
+  public String getResolvedHost(String serviceName) {
+    if (universeDomain != null && universeDomain.isEmpty()) {
+      throw new IllegalArgumentException("Universe Domain cannot be empty");
+    }
+    String resolvedUniverseDomain =
+        universeDomain != null ? universeDomain : Credentials.GOOGLE_DEFAULT_UNIVERSE;
+    // The host value set to DEFAULT_HOST if the user didn't configure a host. If the
+    // user set a host the library uses that value, otherwise, construct the host for the user.
+    // The DEFAULT_HOST value is not a valid host for handwritten libraries.
+    if (!host.equals(DEFAULT_HOST)) {
+      return host;
+    }
+    return "https://www." + serviceName + "." + resolvedUniverseDomain + "/";
+  }
+
+  /**
+   * Temporarily used for BigQuery and Storage Apiary Wrapped Libraries. To be removed in the
+   * future. Returns the host to be used for the rootUrl and output is in the format of:
+   * "https://serviceName.universeDomain/"
+   */
+  @InternalApi
+  public String getResolvedApiaryHost(String serviceName) {
+    String resolvedUniverseDomain =
+        getUniverseDomain() != null ? getUniverseDomain() : Credentials.GOOGLE_DEFAULT_UNIVERSE;
+    return "https://www." + serviceName + "." + resolvedUniverseDomain + "/";
+  }
+
+  /**
+   * Validates that Credentials' Universe Domain matches the user configured Universe Domain.
+   * Currently, this is intended for BigQuery and Storage Apiary Wrapped Libraries
+   */
+  @InternalApi
+  public boolean isValidUniverseDomain() throws IOException {
+    String resolvedUniverseDomain =
+        getUniverseDomain() != null ? getUniverseDomain() : Credentials.GOOGLE_DEFAULT_UNIVERSE;
```

Comment:
It's itchy that getUniverseDomain is called twice. Can you save the returned value in a variable?

Reply:
Sounds good, will update.